File: linkedlist/linked_list_operation/linked_list_operation.py
#Implementation for linked list operations

class node:
    def __init__(self ,data):
        self.data = data
        self.next = None
    
class linked_list:
    def __init__(self):
        self.HEAD = None
    
    ##insert data at begining of linked-list
    def insert_at_begining(self ,data):
        tmp_node = node(data)

        tmp_node.next = self.HEAD
        self.HEAD = tmp_node

    ##insert data at ending of linked-list
    def insert_at_end(self ,data):
        tmp_node = node(data)

        if self.HEAD is None:
            self.HEAD = tmp_node
            return
        
        LAST = self.HEAD
        while( LAST.next):
            LAST = LAST.next
        
        LAST.next = tmp_node
    
    ##search for an element in the linked-list
    def search(self ,key):
        CURR = self.HEAD

        while CURR is not None:
            if CURR.data == key:
                return True
            
            CURR = CURR.next
        
        return False

    ##delete node from specific index
    def delete_element(self ,indx):

        if self.HEAD is None:
            return 
        
        tmp = self.HEAD

        if indx == 0:
            self.HEAD = tmp.next
            tmp = None
            return
        
        for i in range(indx - 1):
            tmp = tmp.next

            if tmp is None:
                return
        if tmp.next is None:
            return
        
        nxt = tmp.next.next
        tmp.next = None

        tmp.next = nxt
    
    ##insert node after specific node
    def insert_after(self ,prev_node ,data):

        if prev_node is None:
            return
        
        tmp_node = node(data)

        tmp_node.next = prev_node.next

        prev_node.next = tmp_node
    
    ##print linked-list
    def print_linked_list(self):
        tmp_node = self.HEAD

        while(tmp_node):
            print(tmp_node.data ),
            tmp_node = tmp_node.next
        
        print("\n")
    
    ##sort the linked_list
    def sort_linked_list(self):
        curr_node = self.HEAD

        while(curr_node):
            next_node = curr_node.next

            while next_node:
                if curr_node.data > next_node.data:
                    ##swap it
                    curr_node.data ,next_node.data = next_node.data ,curr_node.data

                next_node = next_node.next
            
            curr_node = curr_node.next
##
if __name__ == '__main__':

    link_list = linked_list()

    link_list.insert_at_begining(9)
    link_list.insert_at_begining(4)
    link_list.insert_at_begining(2)
    link_list.insert_at_begining(3)

    link_list.print_linked_list()

    link_list.insert_at_end(4)

    link_list.print_linked_list()

    link_list.sort_linked_list()

    link_list.print_linked_list()



    

File: linkedlist/linked_list_operation/linked_list_operation.c
#include <stdio.h>
#include <stdlib.h>

struct node{
    int data;
    struct node *next;
};

//swapping method
void swap(int *num1 ,int *num2){
    int tmp = *num1;
    *num1   = *num2;
    *num2   = tmp;
}

//insert at beginning
void insert_at_begining(struct node **head_ref  ,int data){
    struct node *tmp_node = (struct node*)malloc(sizeof(struct node));

    //insert data into the newly created node
    tmp_node->data = data;
    //set the link of newly created to node pointed by head 
    tmp_node->next = (*head_ref);
    //set the head to the newly created node since the motive is to add the node 
    (*head_ref) = tmp_node;
}

//insert node after specific node 
void insert_after(struct node *prev_node ,int data){
    if(prev_node == NULL){
        return;
    }

    struct node *tmp_node = (struct node*)malloc(sizeof(struct node));

    tmp_node->data = data;
    tmp_node->next = prev_node->next;
    prev_node->next = tmp_node;
}

//
void insert_node_at_last(struct node **head_ref ,int data){

    struct node *tmp_node = (struct node*)malloc(sizeof(struct node));
    tmp_node->data = data;
    tmp_node->next = NULL;

    if(head_ref == NULL){
        *head_ref = tmp_node;
        NULL;
    }

    struct node *last_node = *head_ref;
    
    while(last_node->next != NULL){
        last_node = last_node->next;
    }

    last_node->next = tmp_node;
}

//
void delete_node(struct node **head_ref ,int key){
    struct node *tmp_head_node = *head_ref ,*tmp_prev_node;

    if(tmp_head_node != NULL && tmp_head_node->data  == key){
        *head_ref = tmp_head_node->next;

        free(tmp_head_node);

        return;
    }

    while(tmp_head_node != NULL && tmp_head_node->data != key){
        tmp_prev_node = tmp_head_node;

        tmp_head_node = tmp_head_node->next;
    }

    if(tmp_head_node == NULL){
        return;
    }

    //remove the node
    tmp_prev_node->next = tmp_head_node->next;
    free(tmp_head_node);
}

int search_node(struct node **head_ref ,int data){

    struct node *tmp_node = *head_ref;

    while(tmp_node != NULL){
        if(tmp_node->data == data){
            tmp_node = tmp_node->next;
        }
        
    }
    return 0;
}

void sort_node(struct node *head ){
    struct node *curr_node = head;

    while(curr_node != NULL){
        struct node *next_node = curr_node->next;

        while(next_node != NULL){
            if(curr_node->data > next_node->data){
                swap(&curr_node->data ,&next_node->data);
            }

            next_node = next_node->next;
        }

        curr_node = curr_node->next;
    }
}

void print_linked_list(struct node *node){

    while(node != NULL){
        printf("%d\t" ,node->data);
        node = node->next;
    }

    printf("\n");
}

int main(){
    struct node *head = NULL;

    insert_at_begining(&head ,9);
    insert_at_begining(&head ,43);
    insert_at_begining(&head ,24);
    insert_at_begining(&head ,3);

    print_linked_list(head);

    insert_node_at_last(&head ,4);

    print_linked_list(head);

    delete_node(&head ,0);

    print_linked_list(head);

    insert_after(head->next->next ,9);

    print_linked_list(head);

    sort_node(head);

    print_linked_list(head);
}